src/csharp/Microsoft.Spark/Sql/PicklingUdfWrapper.cs

```diff
@@ -274,24 +390,52 @@ internal object Execute(int splitIndex, object[] input, int[] argOffsets)
     {
+        [NonSerialized]
+        private bool? _sameT1 = null;
+        [NonSerialized]
+        private bool? _sameT2 = null;
+        [NonSerialized]
+        private bool? _sameT3 = null;
+        [NonSerialized]
+        private bool? _sameT4 = null;
+        [NonSerialized]
+        private bool? _sameT5 = null;
+        [NonSerialized]
+        private bool? _sameT6 = null;
+        [NonSerialized]
+        private bool? _sameT7 = null;
+        [NonSerialized]
+        private bool? _sameT8 = null;
+        [NonSerialized]
+        private bool? _sameT9 = null;
+
         private readonly Func<T1, T2, T3, T4, T5, T6, T7, T8, T9, TResult> _func;
 
         internal PicklingUdfWrapper(Func<T1, T2, T3, T4, T5, T6, T7, T8, T9, TResult> func)
         {
             _func = func;
         }
-        internal object Execute(int splitIndex, object[] input, int[] argOffsets)
+        internal object Execute(int _, object[] input, int[] argOffsets)
         {
+            object param1 = input[argOffsets[0]];
+            object param2 = input[argOffsets[1]];
+            object param3 = input[argOffsets[2]];
+            object param4 = input[argOffsets[3]];
+            object param5 = input[argOffsets[4]];
+            object param6 = input[argOffsets[5]];
+            object param7 = input[argOffsets[6]];
+            object param8 = input[argOffsets[7]];
+            object param9 = input[argOffsets[8]];
             return _func(
-                TypeConverter.ConvertTo<T1>(input[argOffsets[0]]),
-                TypeConverter.ConvertTo<T2>(input[argOffsets[1]]),
-                TypeConverter.ConvertTo<T3>(input[argOffsets[2]]),
-                TypeConverter.ConvertTo<T4>(input[argOffsets[3]]),
-                TypeConverter.ConvertTo<T5>(input[argOffsets[4]]),
-                TypeConverter.ConvertTo<T6>(input[argOffsets[5]]),
-                TypeConverter.ConvertTo<T7>(input[argOffsets[6]]),
-                TypeConverter.ConvertTo<T8>(input[argOffsets[7]]),
-                TypeConverter.ConvertTo<T9>(input[argOffsets[8]]));
+                (_sameT1 ??= param1 is T1) ? (T1)param1 : TypeConverter.ConvertTo<T1>(param1),
+                (_sameT2 ??= param2 is T2) ? (T2)param2 : TypeConverter.ConvertTo<T2>(param2),
+                (_sameT3 ??= param3 is T3) ? (T3)param3 : TypeConverter.ConvertTo<T3>(param3),
+                (_sameT4 ??= param4 is T4) ? (T4)param4 : TypeConverter.ConvertTo<T4>(param4),
+                (_sameT5 ??= param5 is T5) ? (T5)param5 : TypeConverter.ConvertTo<T5>(param5),
+                (_sameT6 ??= param6 is T6) ? (T6)param6 : TypeConverter.ConvertTo<T6>(param6),
+                (_sameT7 ??= param7 is T7) ? (T7)param7 : TypeConverter.ConvertTo<T7>(param7),
+                (_sameT8 ??= param8 is T8) ? (T8)param8 : TypeConverter.ConvertTo<T8>(param8),
+                (_sameT9 ??= param9 is T9) ? (T9)param9 : TypeConverter.ConvertTo<T9>(param9));
```

Comment:
I am wondering if this logic can be easily abstracted into a function that is fed the `input` and `argOffsets` and returns the `_func`

Reply:
added